### PR TITLE
Transmission: add support for ports < 1024 (needs systemd >= 229)

### PR DESCRIPTION
**Pull request by davidak**
###### Things done:
- Built on platform(s)
  - [x] NixOS
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

solution from https://github.com/NixOS/nixpkgs/issues/11184

cc @bjornfor @thoughtpolice @wizeman

### CODE REVIEW

File: nixos/modules/services/torrent/transmission.nix [hunk 1]
```diff
@@ -86,6 +86,8 @@ in
         ExecStartPre = ''
           ${pkgs.stdenv.shell} -c "chmod 770 ${homeDir} && mkdir -p ${settingsDir} ${downloadDir} ${incompleteDir} && rm -f ${settingsDir}/settings.json &&   cp -f ${settingsFile} ${settingsDir}/settings.json"
         '';
+        # allow to bind on ports below 1024
+        AmbientCapabilities = "cap_net_bind_service";
```
Comment (joachifm) [inline nixos/modules/services/torrent/transmission.nix]: I think it makes sense to make this conditional on (`cfg.ports < 1024`) and I would also like to see a warning to the user that the daemon will run with elevated capabilities.

**Comment by joachifm** [comment]
The implementation is okay, but I will reiterate what I've said earlier: an nginx reverse proxy is the superior solution.

The RPC/HTTP features of transmission are most likely the least tested wrt. malicious users and I would never expose those parts to the open internet, nor do I think we should make it more convenient to do so.

**Comment by davidak** [comment]
@joachifm yes, that makes sense when exposing to the internet. i'm running it in my local network and don't want it more complicated as needed.

should that get documented in the nixos manual for administrating services?

**Comment by joachifm** [comment]
Ah, I see. I think it would be better if you simply set the ambient capabilites in your configuration.nix. Adding it to the service module gives the impression that it's a supported/recommended mode of operation, which I certainly believe it should not be.

Anyway, I'm happy to defer to the module maintainers in this case, so let's wait for their input.

File: nixos/modules/services/torrent/transmission.nix [hunk 2]
```diff
@@ -82,14 +82,16 @@ in
...
+      serviceConfig = {
+        ExecStartPre = ''
+          ${pkgs.stdenv.shell} -c "chmod 770 ${homeDir} && mkdir -p ${settingsDir} ${downloadDir} ${incompleteDir} && rm -f ${settingsDir}/settings.json &&   cp -f ${settingsFile} ${settingsDir}/settings.json"
+        '';
+        ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f --port ${toString config.services.transmission.port}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        User = "transmission";
+        # NOTE: transmission has an internal umask that also must be set (in settings.json)
+        UMask = "0002";
+      };
```
Comment (bjornfor) [inline nixos/modules/services/torrent/transmission.nix]: -1. This is a unclean clean up. Why do you change ExecStartPre? It used to do "mkdir -p" to ensure working conditions, but now it starts with a "chmod" instead. (Won't that break?) Please don't mix changes like that in "clean up" commits.

**Comment by bjornfor** [comment]
I have no strong feelings about the AmbientCapabilities. But the first clean-up commit I didn't like.

**Comment by bjornfor** [comment]
@davidak: Like @joachifm says, giving transmission more privileges can be done locally, in configuration.nix:

```
systemd.services.transmission.serviceConfig.AmbientCapabilities = "cap_net_bind_service";
```

What do you think about that? Is it still something you think we should have in NixOS, or is it OK to keep it as a local configuration? (The more I think about it, the more I'm agreeing with @joachifm: NixOS services should run with as little permissions as possible.)

**Comment by davidak** [comment]
@bjornfor i totally agree that NixOS services should run with as little permissions as possible.

i changed my config: https://github.com/davidak/nixos-config/commit/44e8f0ed40abbf38faababd6eb4a4c5a2045af1d

**Pull request closed without merging**